### ISSUE TITLE
Excess Namespaces

### ISSUE THREAD

**Issue by DavidRies** [CISecurity/OVALRepo#12]
It looks the elements in the repo have more inline namespaces than necessary:
````xml
<family_test id="oval:org.mitre.oval:tst:99" version="1" comment="the installed operating system is part of the Microsoft Windows family" check_existence="at_least_one_exists" check="only one" xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#independent" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:oval="http://oval.mitre.org/XMLSchema/oval-common-5" xmlns:oval-def="http://oval.mitre.org/XMLSchema/oval-definitions-5">
     <object object_ref="oval:org.mitre.oval:obj:99"/>
     <state state_ref="oval:org.mitre.oval:ste:99"/>
</family_test>
````
could be
````xml
<family_test id="oval:org.mitre.oval:tst:99" version="1" comment="the installed operating system is part of the Microsoft Windows family" check_existence="at_least_one_exists" check="only one" xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#independent">
     <object object_ref="oval:org.mitre.oval:obj:99"/>
     <state state_ref="oval:org.mitre.oval:ste:99"/>
</family_test>
````

**Comment by GunnarEngelbach** [comment]
Good point, but one must be careful with this.  Object and state and not defined in the independent schema, and other elements might need to include other schema that define things like xsi:nil.

When multiple elements are combined into a single XML document the necessary schema will likely be defined in the document root, so it will be a consideration in the scripts that build these documents to also be sure all the necessary schema references are present.

**Comment by DavidRies** [comment]
Currently, the build scripts assume that these fragments can be concatenated together under an <oval_definitions> element that looks like this:
```xml
<oval_definitions 
   xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5" 
   xmlns:oval="http://oval.mitre.org/XMLSchema/oval-common-5" 
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
   xsi:schemaLocation="http://oval.mitre.org/XMLSchema/oval-common-5 oval-common-schema.xsd http://oval.mitre.org/XMLSchema/oval-definitions-5 oval-definitions-schema.xsd">
```
That means that fragments will have to contain the namespaces they use inline (with the exception of the ones declared above).

Does that work?

And, I still don't understand why we need all those extra inline namespaces.

**Comment by GunnarEngelbach** [comment]
The extra inline namespaces are not necessary.  The question is, for any particular XML file, which ones are?

Same problem when combining multiple files into a single one.  It probably won't need a declaration of all of the namespaces, but which ones it does require depends on which content is added to it.

It's just something to keep in mind when creating a merged document:  every namespace needed by a child component needs to be verified as being declared in the parent document.  It might also be necessary to explicitly remove the NS declaration from the child element.

**Comment by DavidRies** [comment]
My thought is that each fragment should include the namespaces that are used in that fragment and--ideally--no others. Doesn't that make the fragments independent from the parent? That way we don't need to worry about reconciling parent NS declarations with fragments, etc.

**Comment by GunnarEngelbach** [comment]
Yes, that would be the ideal.  And my point is that the example you showed of a cleaned up file is incomplete.  For that example, it should also reference the oval-definitions-5 schema.  And other elements can require more schema.

Simply including all possible schema in the root element makes coding very easy, which I'm sure is why it's done like that most often.  Programmatically determining the minimal complement of schema necessary is a bit more complicated.

It's going to take something like a table lookup to be able to do that.  Which also means maintaining that reference as the schema change.  But that's not an entirely bad thing, because that same reference could be used to determine the minimal schema version needed for any element -- which would be considerably more efficient than iterating through schema validations until one failed.

The problem is, as OtherDavid pointed out under another item, the same element name is used in more than one schema.  To disambiguate you'd have to already know the correct schema.
